File: python/src/pi_mono/agent/proxy.py
import json
import time
from typing import Any, Union, cast, Literal

import httpx
from pi_mono.ai.types import (
    AssistantMessage,
    AssistantMessageEvent,
    Context,
    Model,
    SimpleStreamOptions,
    ToolCall,
)
from pi_mono.utils.event_stream import EventStream
from pi_mono.utils.json_parse import parse_streaming_json


class ProxyMessageEventStream(EventStream[AssistantMessageEvent, AssistantMessage]):
    def __init__(self) -> None:
        super().__init__(
            is_complete=lambda event: event["type"] in ("done", "error"),
            extract_result=self._extract_result,
        )

    def _extract_result(self, event: AssistantMessageEvent) -> AssistantMessage:
        if event["type"] == "done":
            return event["message"]
        elif event["type"] == "error":
            return event["error"]
        raise ValueError("Unexpected event type for final result")


ProxyAssistantMessageEvent = Union[
    dict[str, Any],  # general dictionary matching ProxyAssistantMessageEvent
]


class ProxyStreamOptions(SimpleStreamOptions, total=False):
    authToken: str
    proxyUrl: str


def build_proxy_request_options(options: ProxyStreamOptions) -> dict[str, Any]:
    serialized: dict[str, Any] = {}
    for key in (
        "temperature",
        "maxTokens",
        "reasoning",
        "cacheRetention",
        "sessionId",
        "headers",
        "metadata",
        "transport",
        "thinkingBudgets",
        "maxRetryDelayMs",
    ):
        if key in options:
            serialized[key] = options[key]  # type: ignore
    return serialized


def ensure_content_index(content_list: list[Any], index: int, default_val: Any = None) -> None:
    while len(content_list) <= index:
        content_list.append(default_val)


def process_proxy_event(
    proxy_event: dict[str, Any],
    partial: AssistantMessage,
) -> AssistantMessageEvent | None:
    event_type = proxy_event.get("type")

    if event_type == "start":
        return {"type": "start", "partial": partial}

    elif event_type == "text_start":
        content_index = proxy_event["contentIndex"]
        ensure_content_index(partial["content"], content_index)
        partial["content"][content_index] = {"type": "text", "text": ""}
        return {"type": "text_start", "contentIndex": content_index, "partial": partial}

    elif event_type == "text_delta":
        content_index = proxy_event["contentIndex"]
        ensure_content_index(partial["content"], content_index)
        content = partial["content"][content_index]
        if content and content.get("type") == "text":
            content["text"] += proxy_event["delta"]
            return {
                "type": "text_delta",
                "contentIndex": content_index,
                "delta": proxy_event["delta"],
                "partial": partial,
            }
        raise ValueError("Received text_delta for non-text content")

    elif event_type == "text_end":
        content_index = proxy_event["contentIndex"]
        ensure_content_index(partial["content"], content_index)
        content = partial["content"][content_index]
        if content and content.get("type") == "text":
            if "contentSignature" in proxy_event:
                content["textSignature"] = proxy_event["contentSignature"]
            return {
                "type": "text_end",
                "contentIndex": content_index,
                "content": content["text"],
                "partial": partial,
            }
        raise ValueError("Received text_end for non-text content")

    elif event_type == "thinking_start":
        content_index = proxy_event["contentIndex"]
        ensure_content_index(partial["content"], content_index)
        partial["content"][content_index] = {"type": "thinking", "thinking": ""}
        return {"type": "thinking_start", "contentIndex": content_index, "partial": partial}

    elif event_type == "thinking_delta":
        content_index = proxy_event["contentIndex"]
        ensure_content_index(partial["content"], content_index)
        content = partial["content"][content_index]
        if content and content.get("type") == "thinking":
            content["thinking"] += proxy_event["delta"]
            return {
                "type": "thinking_delta",
                "contentIndex": content_index,
                "delta": proxy_event["delta"],
                "partial": partial,
            }
        raise ValueError("Received thinking_delta for non-thinking content")

    elif event_type == "thinking_end":
        content_index = proxy_event["contentIndex"]
        ensure_content_index(partial["content"], content_index)
        content = partial["content"][content_index]
        if content and content.get("type") == "thinking":
            if "contentSignature" in proxy_event:
                content["thinkingSignature"] = proxy_event["contentSignature"]
            return {
                "type": "thinking_end",
                "contentIndex": content_index,
                "content": content["thinking"],
                "partial": partial,
            }
        raise ValueError("Received thinking_end for non-thinking content")

    elif event_type == "toolcall_start":
        content_index = proxy_event["contentIndex"]
        ensure_content_index(partial["content"], content_index)
        tool_call_obj = cast(
            ToolCall,
            {
                "type": "toolCall",
                "id": proxy_event["id"],
                "name": proxy_event["toolName"],
                "arguments": {},
                "partialJson": "",
            },
        )
        partial["content"][content_index] = tool_call_obj
        return {"type": "toolcall_start", "contentIndex": content_index, "partial": partial}

    elif event_type == "toolcall_delta":
        content_index = proxy_event["contentIndex"]
        ensure_content_index(partial["content"], content_index)
        content = partial["content"][content_index]
        if content and content.get("type") == "toolCall":
            dict_content = cast(dict[str, Any], content)
            dict_content["partialJson"] = dict_content.get("partialJson", "") + proxy_event["delta"]
            dict_content["arguments"] = parse_streaming_json(dict_content["partialJson"]) or {}
            return {
                "type": "toolcall_delta",
                "contentIndex": content_index,
                "delta": proxy_event["delta"],
                "partial": partial,
            }
        raise ValueError("Received toolcall_delta for non-toolCall content")

    elif event_type == "toolcall_end":
        content_index = proxy_event["contentIndex"]
        ensure_content_index(partial["content"], content_index)
        content = partial["content"][content_index]
        if content and content.get("type") == "toolCall":
            dict_content = cast(dict[str, Any], content)
            if "partialJson" in dict_content:
                del dict_content["partialJson"]
            return {
                "type": "toolcall_end",
                "contentIndex": content_index,
                "toolCall": content,
                "partial": partial,
            }
        return None

    elif event_type == "done":
        partial["stopReason"] = proxy_event["reason"]
        partial["usage"] = proxy_event["usage"]
        return {"type": "done", "reason": proxy_event["reason"], "message": partial}

    elif event_type == "error":
        partial["stopReason"] = proxy_event["reason"]
        if "errorMessage" in proxy_event:
            partial["errorMessage"] = proxy_event["errorMessage"]
        partial["usage"] = proxy_event["usage"]
        return {"type": "error", "reason": proxy_event["reason"], "error": partial}

    return None


def stream_proxy(
    model: Model,
    context: Context,
    options: ProxyStreamOptions,
) -> ProxyMessageEventStream:
    stream = ProxyMessageEventStream()

    async def run() -> None:
        partial: AssistantMessage = {
            "role": "assistant",
            "stopReason": "stop",
            "content": [],
            "api": model.get("api", "unknown"),
            "provider": model.get("provider", "unknown"),
            "model": model.get("id", "unknown"),
            "usage": {
                "input": 0,
                "output": 0,
                "cacheRead": 0,
                "cacheWrite": 0,
                "totalTokens": 0,
                "cost": {"input": 0, "output": 0, "cacheRead": 0, "cacheWrite": 0, "total": 0},
            },
            "timestamp": int(time.time() * 1000),
        }

        signal = options.get("signal")
        client = httpx.AsyncClient()

        # Handle cancellation
        abort_called = False

        def abort_handler() -> None:
            nonlocal abort_called
            abort_called = True

        if signal:
            signal.add_event_listener("abort", abort_handler, once=True)

        try:
            if signal and signal.aborted:
                raise RuntimeError("Request aborted by user")

            async with client.stream(
                "POST",
                f"{options['proxyUrl']}/api/stream",
                headers={
                    "Authorization": f"Bearer {options['authToken']}",
                    "Content-Type": "application/json",
                },
                json={
                    "model": model,
                    "context": context,
                    "options": build_proxy_request_options(options),
                },
                timeout=None,
            ) as response:
                if abort_called or (signal and signal.aborted):
                    raise RuntimeError("Request aborted by user")

                if response.status_code != 200:
                    error_msg = f"Proxy error: {response.status_code} {response.reason_phrase}"
                    try:
                        error_data = response.json()
                        if isinstance(error_data, dict) and "error" in error_data:
                            error_msg = f"Proxy error: {error_data['error']}"
                    except Exception:
                        pass
                    raise RuntimeError(error_msg)

                async for line in response.aiter_lines():
                    if abort_called or (signal and signal.aborted):
                        raise RuntimeError("Request aborted by user")

                    if line.startswith("data: "):
                        data = line[6:].strip()
                        if data:
                            proxy_event = json.loads(data)
                            event = process_proxy_event(proxy_event, partial)
                            if event:
                                stream.push(event)

            if signal and signal.aborted:
                raise RuntimeError("Request aborted by user")

            stream.end()
        except Exception as error:
            error_message = str(error)
            reason: Literal["aborted", "error"] = (
                "aborted" if (signal and signal.aborted) else "error"
            )
            partial["stopReason"] = reason
            partial["errorMessage"] = error_message
            stream.push(
                {
                    "type": "error",
                    "reason": reason,
                    "error": partial,
                }
            )
            stream.end()
        finally:
            await client.aclose()
            if signal:
                signal.remove_event_listener("abort", abort_handler)

    import asyncio

    asyncio.create_task(run())
    return stream

File: packages/agent/src/proxy.ts
/**
 * 通过服务端中转 LLM 调用的代理 stream 函数。
 * 服务端管理认证并将请求代理转发给 LLM provider。
 */

// 导入 JSON 解析工具的内部模块
import {
	type AssistantMessage,
	type AssistantMessageEvent,
	type Context,
	EventStream,
	type Model,
	parseStreamingJson,
	type SimpleStreamOptions,
	type StopReason,
	type ToolCall,
} from "@earendil-works/pi-ai";

// 与 ProxyMessageEventStream 匹配的 stream 类，在事件流结束时提取最终的 AssistantMessage。
class ProxyMessageEventStream extends EventStream<AssistantMessageEvent, AssistantMessage> {
	constructor() {
		super(
			(event) => event.type === "done" || event.type === "error",
			(event) => {
				if (event.type === "done") return event.message;
				if (event.type === "error") return event.error;
				throw new Error("Unexpected event type");
			},
		);
	}
}

/**
 * 代理事件类型 —— 服务端发送时移除了 partial 字段以节省带宽。
 */
export type ProxyAssistantMessageEvent =
	| { type: "start" }
	| { type: "text_start"; contentIndex: number }
	| { type: "text_delta"; contentIndex: number; delta: string }
	| { type: "text_end"; contentIndex: number; contentSignature?: string }
	| { type: "thinking_start"; contentIndex: number }
	| { type: "thinking_delta"; contentIndex: number; delta: string }
	| { type: "thinking_end"; contentIndex: number; contentSignature?: string }
	| { type: "toolcall_start"; contentIndex: number; id: string; toolName: string }
	| { type: "toolcall_delta"; contentIndex: number; delta: string }
	| { type: "toolcall_end"; contentIndex: number }
	| {
			type: "done";
			reason: Extract<StopReason, "stop" | "length" | "toolUse">;
			usage: AssistantMessage["usage"];
	  }
	| {
			type: "error";
			reason: Extract<StopReason, "aborted" | "error">;
			errorMessage?: string;
			usage: AssistantMessage["usage"];
	  };

type ProxySerializableStreamOptions = Pick<
	SimpleStreamOptions,
	| "temperature"
	| "maxTokens"
	| "reasoning"
	| "cacheRetention"
	| "sessionId"
	| "headers"
	| "metadata"
	| "transport"
	| "thinkingBudgets"
	| "maxRetryDelayMs"
>;

export interface ProxyStreamOptions extends ProxySerializableStreamOptions {
	/** 代理请求的本地中止信号 */
	signal?: AbortSignal;
	/** 代理服务器的认证令牌 */
	authToken: string;
	/** 代理服务器 URL（例如 "https://genai.example.com"） */
	proxyUrl: string;
}

/**
 * 从完整的 ProxyStreamOptions 中提取可序列化的 stream 选项，用于发送给代理服务器。
 */
function buildProxyRequestOptions(options: ProxyStreamOptions): ProxySerializableStreamOptions {
	return {
		temperature: options.temperature,
		maxTokens: options.maxTokens,
		reasoning: options.reasoning,
		cacheRetention: options.cacheRetention,
		sessionId: options.sessionId,
		headers: options.headers,
		metadata: options.metadata,
		transport: options.transport,
		thinkingBudgets: options.thinkingBudgets,
		maxRetryDelayMs: options.maxRetryDelayMs,
	};
}

/**
 * 通过代理服务器中转 LLM 调用的 stream 函数。
 * 服务端移除 delta 事件中的 partial 字段以节省带宽，客户端负责重建完整的 partial message。
 *
 * 在创建需要通过代理的 Agent 时，将此函数作为 `streamFn` 选项传入。
 */
export function streamProxy(model: Model<any>, context: Context, options: ProxyStreamOptions): ProxyMessageEventStream {
	const stream = new ProxyMessageEventStream();

	(async () => {
		/** 从事件流逐步构建的 partial assistant message。 */
		const partial: AssistantMessage = {
			role: "assistant",
			stopReason: "pending",
			content: [],
			api: model.api,
			provider: model.provider,
			model: model.id,
			usage: {
				input: 0,
				output: 0,
				cacheRead: 0,
				cacheWrite: 0,
				totalTokens: 0,
				cost: { input: 0, output: 0, cacheRead: 0, cacheWrite: 0, total: 0 },
			},
			timestamp: Date.now(),
		};

		let reader: ReadableStreamDefaultReader<Uint8Array> | undefined;

		const abortHandler = () => {
			if (reader) {
				reader.cancel("Request aborted by user").catch(() => {});
			}
		};

		if (options.signal) {
			options.signal.addEventListener("abort", abortHandler);
		}

		try {
			const response = await fetch(`${options.proxyUrl}/api/stream`, {
				method: "POST",
				headers: {
					Authorization: `Bearer ${options.authToken}`,
					"Content-Type": "application/json",
				},
				body: JSON.stringify({
					model,
					context,
					options: buildProxyRequestOptions(options),
				}),
				signal: options.signal,
			});

			if (!response.ok) {
				let errorMessage = `Proxy error: ${response.status} ${response.statusText}`;
				try {
					const errorData = (await response.json()) as { error?: string };
					if (errorData.error) {
						errorMessage = `Proxy error: ${errorData.error}`;
					}
				} catch {
					// 无法解析错误响应
				}
				throw new Error(errorMessage);
			}

			reader = response.body!.getReader();
			const decoder = new TextDecoder();
			let buffer = "";

			while (true) {
				const { done, value } = await reader.read();
				if (done) break;

				if (options.signal?.aborted) {
					throw new Error("Request aborted by user");
				}

				buffer += decoder.decode(value, { stream: true });
				const lines = buffer.split("\n");
				buffer = lines.pop() || "";

				for (const line of lines) {
					if (line.startsWith("data: ")) {
						const data = line.slice(6).trim();
						if (data) {
							const proxyEvent = JSON.parse(data) as ProxyAssistantMessageEvent;
							const event = processProxyEvent(proxyEvent, partial);
							if (event) {
								stream.push(event);
							}
						}
					}
				}
			}

			if (options.signal?.aborted) {
				throw new Error("Request aborted by user");
			}

			stream.end();
		} catch (error) {
			const errorMessage = error instanceof Error ? error.message : String(error);
			const reason = options.signal?.aborted ? "aborted" : "error";
			partial.stopReason = reason;
			partial.errorMessage = errorMessage;
			stream.push({
				type: "error",
				reason,
				error: partial,
			});
			stream.end();
		} finally {
			if (options.signal) {
				options.signal.removeEventListener("abort", abortHandler);
			}
		}
	})();

	return stream;
}

/**
 * 处理代理事件并更新 partial 消息。
 */
function processProxyEvent(
	proxyEvent: ProxyAssistantMessageEvent,
	partial: AssistantMessage,
): AssistantMessageEvent | undefined {
	switch (proxyEvent.type) {
		case "start":
			return { type: "start", partial };

		case "text_start":
			partial.content[proxyEvent.contentIndex] = { type: "text", text: "" };
			return { type: "text_start", contentIndex: proxyEvent.contentIndex, partial };

		case "text_delta": {
			const content = partial.content[proxyEvent.contentIndex];
			if (content?.type === "text") {
				content.text += proxyEvent.delta;
				return {
					type: "text_delta",
					contentIndex: proxyEvent.contentIndex,
					delta: proxyEvent.delta,
					partial,
				};
			}
			throw new Error("Received text_delta for non-text content");
		}

		case "text_end": {
			const content = partial.content[proxyEvent.contentIndex];
			if (content?.type === "text") {
				content.textSignature = proxyEvent.contentSignature;
				return {
					type: "text_end",
					contentIndex: proxyEvent.contentIndex,
					content: content.text,
					partial,
				};
			}
			throw new Error("Received text_end for non-text content");
		}

		case "thinking_start":
			partial.content[proxyEvent.contentIndex] = { type: "thinking", thinking: "" };
			return { type: "thinking_start", contentIndex: proxyEvent.contentIndex, partial };

		case "thinking_delta": {
			const content = partial.content[proxyEvent.contentIndex];
			if (content?.type === "thinking") {
				content.thinking += proxyEvent.delta;
				return {
					type: "thinking_delta",
					contentIndex: proxyEvent.contentIndex,
					delta: proxyEvent.delta,
					partial,
				};
			}
			throw new Error("Received thinking_delta for non-thinking content");
		}

		case "thinking_end": {
			const content = partial.content[proxyEvent.contentIndex];
			if (content?.type === "thinking") {
				content.thinkingSignature = proxyEvent.contentSignature;
				return {
					type: "thinking_end",
					contentIndex: proxyEvent.contentIndex,
					content: content.thinking,
					partial,
				};
			}
			throw new Error("Received thinking_end for non-thinking content");
		}

		case "toolcall_start":
			partial.content[proxyEvent.contentIndex] = {
				type: "toolCall",
				id: proxyEvent.id,
				name: proxyEvent.toolName,
				arguments: {},
				partialJson: "",
			} satisfies ToolCall & { partialJson: string } as ToolCall;
			return { type: "toolcall_start", contentIndex: proxyEvent.contentIndex, partial };

		case "toolcall_delta": {
			const content = partial.content[proxyEvent.contentIndex];
			if (content?.type === "toolCall") {
				(content as any).partialJson += proxyEvent.delta;
				content.arguments = parseStreamingJson((content as any).partialJson) || {};
				partial.content[proxyEvent.contentIndex] = { ...content }; // 触发响应式更新
				return {
					type: "toolcall_delta",
					contentIndex: proxyEvent.contentIndex,
					delta: proxyEvent.delta,
					partial,
				};
			}
			throw new Error("Received toolcall_delta for non-toolCall content");
		}

		case "toolcall_end": {
			const content = partial.content[proxyEvent.contentIndex];
			if (content?.type === "toolCall") {
				delete (content as any).partialJson;
				return {
					type: "toolcall_end",
					contentIndex: proxyEvent.contentIndex,
					toolCall: content,
					partial,
				};
			}
			return undefined;
		}

		case "done":
			partial.stopReason = proxyEvent.reason;
			partial.usage = proxyEvent.usage;
			return { type: "done", reason: proxyEvent.reason, message: partial };

		case "error":
			partial.stopReason = proxyEvent.reason;
			partial.errorMessage = proxyEvent.errorMessage;
			partial.usage = proxyEvent.usage;
			return { type: "error", reason: proxyEvent.reason, error: partial };

		default: {
			const _exhaustiveCheck: never = proxyEvent;
			console.warn(`Unhandled proxy event type: ${(proxyEvent as any).type}`);
			return undefined;
		}
	}
}
